### PR TITLE
GRF-671 : Create Category Version for Enriched Category

### PR DESCRIPTION
**Pull request by antoinetrouve**
**Description (for Contributor and Core Developer)**

**Update Version mechanism** as expected by this [US](https://akeneo.atlassian.net/jira/software/projects/GRF/boards/301?selectedIssue=GRF-671).

**Risk analysis** : No risk as we created a new internal endpoint and the legacy mechanism remains unchanged.

We created a **new service API in Versionning bundle** to create a Version entity for a Category model free from Doctrine. We keep the original process (changeset, save in database, get author by the original event).
- New service : https://github.com/akeneo/pim-community-dev/blob/2ac71bfe2027bef64feec335f78645ceed9ca19e/src/Akeneo/Tool/Bundle/VersioningBundle/ServiceApi/VersionBuilder.php#L33

BuildVersion original process : 
- https://github.com/akeneo/pim-community-dev/blob/2ac71bfe2027bef64feec335f78645ceed9ca19e/src/Akeneo/Tool/Bundle/VersioningBundle/Builder/VersionBuilder.php#L46
- https://github.com/akeneo/pim-community-dev/blob/a03b9ada8e04e8b37aa90c10f21fd84d0daac66c/src/Akeneo/Tool/Bundle/VersioningBundle/Manager/VersionManager.php#L114

**Some important difference** :
- The service do not manage the buildVersion by UUID for now.
- We just need a snapshot, a resourceId and a resourceName to buildVersion.
      ->  We do not expect the versionnable object as a parameter.
      ->  resourceId is still nedded to get a previous version and build the changeset
      ->  resourceName is needeed because we change the versionnable object with a new Model Category which is detached from Doctrine. So we need to keep history with the Category doctrine model, the original versionnable object. See [here](https://github.com/akeneo/pim-community-dev/blob/2ac71bfe2027bef64feec335f78645ceed9ca19e/src/Akeneo/Tool/Bundle/VersioningBundle/Builder/VersionBuilder.php#L56).
      - The snapshot to avoid coupling and avoid using the normalization process [here](https://github.com/akeneo/pim-community-dev/blob/2ac71bfe2027bef64feec335f78645ceed9ca19e/src/Akeneo/Tool/Bundle/VersioningBundle/Builder/VersionBuilder.php#L56). and [here](https://github.com/akeneo/pim-community-dev/blob/2ac71bfe2027bef64feec335f78645ceed9ca19e/src/Akeneo/Category/back/Infrastructure/EventSubscriber/UpdateCategoryVersionSubscriber.php#L16). We add or own "normalization" [here](https://github.com/akeneo/pim-community-dev/blob/2ac71bfe2027bef64feec335f78645ceed9ca19e/src/Akeneo/Category/back/Infrastructure/Builder/CategoryVersionBuilder.php#L14)

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Category/back/Infrastructure/EventSubscriber/UpdateCategoryVersionSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            CategoryUpdatedEvent::class => 'UpdateCategoryVersion',
+        ];
+    }
+
+    public function UpdateCategoryVersion(CategoryUpdatedEvent $event): void
```
Comment (olivierpontier) [inline src/Akeneo/Category/back/Infrastructure/EventSubscriber/UpdateCategoryVersionSubscriber.php:31]:
```suggestion
    public function updateCategoryVersion(CategoryUpdatedEvent $event): void
```

File: src/Akeneo/Tool/Bundle/VersioningBundle/tests/integration/ServiceApi/VersionBuilderIntegration.php [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    private ObjectManager $objectManager;
+    private LegacyVersionBuilder $legacyVersionBuilder;
+
+    protected function setUp(): void
+    {
+        parent::setUp();
+        $this->versionFactory = $this->get('pim_versioning.factory.version');
+        $this->objectManager = $this->get('doctrine.orm.default_entity_manager');
+        $this->legacyVersionBuilder = $this->get('pim_versioning.builder.version');
+
```
Comment (amigaire) [inline src/Akeneo/Tool/Bundle/VersioningBundle/tests/integration/ServiceApi/VersionBuilderIntegration.php:34]:
```suggestion
```

File: src/Akeneo/Category/back/Domain/ValueObject/Version/CategoryVersion.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @phpstan-type Snapshot array{
+ *     code: string,
+ *     parent: string,
+ *     updated: string
+ * }&array<string, string>
+ */
+class CategoryVersion
+{
+    public const CATEGORY_VERSION_RESOURCE_NAME = "Akeneo\Category\Infrastructure\Component\Model\Category";
```
Comment (amigaire) [inline src/Akeneo/Category/back/Domain/ValueObject/Version/CategoryVersion.php:19]: Is it the string which was involved in an SLA (I can't remember exactly the bug)?
I thought we wanted to find another way, more stable than class path

**Review by amigaire — APPROVED** [review]

File: src/Akeneo/Category/back/Domain/ValueObject/Version/CategoryVersion.php (hunk 3 above)
Comment (amigaire) [reply src/Akeneo/Category/back/Domain/ValueObject/Version/CategoryVersion.php:19]: I just read the PR's description, my bad

Comment (antoinetrouve) [reply src/Akeneo/Category/back/Domain/ValueObject/Version/CategoryVersion.php:19]: Yes we have to keep the original class name to keep history version with  the Category Doctrine model. We will change the original name by the another one when we will be completely free from Doctrine AND we host in the Category BC the Category version process (means Build Version model, compute the changeset, save in db).

For now we have to manage two model (Dcotrine and the new one) and the versionning process is hosted by the VersionningBundle. We keep control over the data needed to instantiate the version object.

**Review by etiennekerbastard — APPROVED** [review]

**Pull request merged**